external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+    private transient ScheduledExecutorService commitOffsetsTask;
+    private transient Lock ackCommitLock;
+    private transient volatile boolean commit;
+    private transient Map<org.apache.storm.kafka.spout.MessageId, Values> emittedTuples;           // Keeps a list of emitted tuples that are pending being acked or failed
+    private transient Map<TopicPartition, Set<org.apache.storm.kafka.spout.MessageId>> failed;     // failed tuples. They stay in this list until success or max retries is reached
+    private transient Map<TopicPartition, OffsetEntry> acked;         // emitted tuples that were successfully acked. These tuples will be committed by the commitOffsetsTask or on consumer rebalance
+    private transient Set<org.apache.storm.kafka.spout.MessageId> blackList;                       // all the tuples that are in traffic when the rebalance occurs will be added to black list to be disregarded when they are either acked or failed
+    private transient int maxRetries;
+
+    public KafkaSpout(org.apache.storm.kafka.spout.KafkaSpoutConfig<K,V> kafkaSpoutConfig, org.apache.storm.kafka.spout.KafkaSpoutStream kafkaSpoutStream, org.apache.storm.kafka.spout.KafkaTupleBuilder<K,V> tupleBuilder) {
+        this.kafkaSpoutConfig = kafkaSpoutConfig;                 // Pass in configuration
+        this.kafkaSpoutStream = kafkaSpoutStream;
+        this.tupleBuilder = tupleBuilder;
+    }
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        // Spout internals
+        this.conf = conf;
+        this.context = context;
+        this.collector = collector;
+
+        // Bookkeeping objects
+        emittedTuples = new HashMap<>();
+        failed = new HashMap<>();
+        acked = new HashMap<>();
+        blackList = new HashSet<>();
+        ackCommitLock = new ReentrantLock();
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+
+        // Kafka consumer
+        kafkaConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps(),
+                kafkaSpoutConfig.getKeyDeserializer(), kafkaSpoutConfig.getValueDeserializer());
+        kafkaConsumer.subscribe(kafkaSpoutConfig.getSubscribedTopics(), new KafkaSpoutConsumerRebalanceListener());
+
+        // Create commit offsets task
+        if (!kafkaSpoutConfig.isConsumerAutoCommitMode()) {     // If it is auto commit, no need to commit offsets manually
+            createCommitOffsetsTask();
+        }
+    }
+
+    // ======== Commit Offsets Task =======
+
+    private void createCommitOffsetsTask() {
+        commitOffsetsTask = Executors.newSingleThreadScheduledExecutor(commitOffsetsThreadFactory());
+        commitOffsetsTask.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                commit = true;
+            }
+        }, 1000, kafkaSpoutConfig.getOffsetsCommitFreqMs(), TimeUnit.MILLISECONDS);
```

Comment:
@revans2 it may be possible but depends on what we want to do. My initial implementation used this thread to do the commits to Kafka. However that causes ConcurrentModificationException on the Kafka side, as it is single threaded. The immediate fix was to just create this volatile variable and leave it as is. This was meant to be temporary all along, an just pushed it like this for an early review attending to the urgency of the patch.

Nevertheless, we should discuss if we would like to consider in the future using the Kafka option commitAsync, which receives as parameter OffsetCommitCallback, which could be used to manage the bookkeeping state. I suppose even if we consider this it won't be in this patch... but I wonder if this piece of functionality would helpful for that scenario, if we ever want to consider it.

I am referring to this comment as C_1 in another comment bellow.